src/Bicep.Cli/Services/TemplateEvaluatorService.cs

```diff
@@ -0,0 +1,208 @@
+            evaluationContext.EvaluateFunction = (FunctionExpression functionExpression, FunctionArgument[] parameters, TemplateErrorAdditionalInfo additionalInfo) =>
+            {
+                if (functionExpression.Function.StartsWithOrdinalInsensitively(LanguageConstants.ListFunctionPrefix) && config.OnListFunc is not null)
+                {
+                    return config.OnListFunc(
+                        functionExpression.Function,
+                        parameters[0].Token.ToString(),
+                        parameters[1].Token.ToString(),
+                        parameters.Length > 2 ? parameters[2].Token : null);
+                }
+
+                if (functionExpression.Function.EqualsOrdinalInsensitively("reference"))
+                {
+                    var resourceId = parameters[0].Token.ToString();
+                    var apiVersion = parameters.Length > 1 ? parameters[1].Token.ToString() : null;
+                    var fullBody = parameters.Length > 2 ? parameters[2].Token.ToString().EqualsOrdinalInsensitively("Full") : false;
+
+                    if (apiVersion is not null && config.OnReferenceFunc is not null)
+                    {
+                        return config.OnReferenceFunc(resourceId, apiVersion, fullBody);
+                    }
+
+                    if (resourceLookup.TryGetValue(resourceId, out var foundResource) &&
+                        (apiVersion is null || StringComparer.OrdinalIgnoreCase.Equals(apiVersion, foundResource.ApiVersion.Value)))
+                    {
+                        return fullBody ? foundResource.ToJToken() : foundResource.Properties.ToJToken();
+                    }
+                }
+
+                var value = defaultEvaluateFunction(functionExpression, parameters, additionalInfo);
+                return value;
+            };
+
+            for (int i = 0; i < template.Resources.Length; i++)
+            {
+                var resource = template.Resources[i];
+
+                if (resource.Properties is not null)
+                {
+                    var skipEvaluationPaths = new InsensitiveHashSet();
+                    if (resource.Type.Value.EqualsOrdinalInsensitively("Microsoft.Resources/deployments"))
+                    {
+                        skipEvaluationPaths.Add("template");
+                    };
+
+                    resource.Properties.Value = ExpressionsEngine.EvaluateLanguageExpressionsRecursive(
+                        root: resource.Properties.Value,
+                        evaluationContext: evaluationContext,
+                        skipEvaluationPaths: skipEvaluationPaths);
+                }
+            }
+
+            if (template.Outputs is not null && template.Outputs.Count > 0)
+            {
+                foreach (var outputKey in template.Outputs.Keys.ToList())
+                {
+                    template.Outputs[outputKey].Value.Value = ExpressionsEngine.EvaluateLanguageExpressionsRecursive(
```

Comment:
We need to call `EvaluateLanguageExpressionsOptimistically` instead. `EvaluateLanguageExpressionsRecursive` will throw if an output value depends on a runtime resource property.